### ISSUE TITLE
Differentiate between Adam and AdamW Optimizer

### ISSUE THREAD

**Issue by kamalesh0406** [google/flax#753]
I have noticed that the code for the Adam Optimizer is actually implementing the weight decay in a manner similar to the one proposed in the [AdamW](https://arxiv.org/pdf/1711.05101.pdf) paper. Should we write it as two different methods if someone wanted to test their models with Adam Optimizer rather than AdamW?

**Comment by 8bitmp3** [comment]
+1 @kamalesh0406 

DeepMind's Optax library (JAX) also includes Adam and AdamW (https://github.com/deepmind/optax/blob/master/optax/__init__.py) and PyTorch has `torch.optim.Adam` and `torch.optim.AdamW`.

**Comment by rwbfd** [comment]
If I am correct, setting weight decay to 0 you will get back to Adam. It has been a while, I might be wrong

**Comment by rwbfd** [comment]
I think you should use an option for whether to degrade to SGD. Then it should be equivalent.

**Comment by rwightman** [comment]
@kamalesh0406 FWIW you get normal Adam behaviour if you set weight_decay arg to the optimizer to 0 (the default) and apply L2 penalty (if desired) to the loss.

I think a pattern where weight_decay applied in the optimizer == decoupled decay (AdamW/SGDW style), otherwise apply L2 (or other) penalty to loss (TF/Keras style) would make sense. You'd have less duplication of optimizer code like you get with PyTorch where you have almost full duplication of code minus the two lines were WD is applied between W/non-W variants. However, the current Flax impl of Momentum already breaks that as it's weight_decay arg applies L2 in the optimizer.

**Comment by levskaya** [comment]
Closing this issue as we're moving to recommend the use of [optax](https://github.com/deepmind/optax) rather than internal optimizers (which does separate out weight decay from adam update).

**Comment by 8bitmp3** [comment]
> Closing this issue as we're moving to recommend the use of [optax](https://github.com/deepmind/optax) rather than internal optimizers (which does separate out weight decay from adam update).

Thanks @levskaya 👍 

@kamalesh0406 fyi ⬇️ 

> DeepMind's Optax library (JAX) also includes Adam and AdamW (https://github.com/deepmind/optax/blob/master/optax/__init__.py)

Optax repo: https://github.com/deepmind/optax
Optax docs: https://optax.readthedocs.io/en/latest/

Hope this helps
